File: frontend/src/renderer/lib/block-actions.ts
import { blockDisplay, type SessionBlock } from "./session-block";

export type BlockActionKind = "copy_block" | "copy_command" | "copy_output" | "rerun" | "rewind";

export type BlockActionContext = {
	mode: "tui" | "chat";
	capabilities: readonly string[];
	canSend: boolean;
	turnInFlight: boolean;
	rollbackableTurnIds: readonly string[];
};

export type BlockAction = { kind: BlockActionKind; payload?: string; turnId?: string };

export function blockActionsFor(block: SessionBlock, ctx: BlockActionContext): BlockAction[] {
	const actions: BlockAction[] = [{ kind: "copy_block", payload: blockCopyText(block) }];
	if (block.detail?.type === "shell" && block.detail.command) actions.push({ kind: "copy_command", payload: block.detail.command });
	if (block.detail?.type === "shell" && block.detail.output) actions.push({ kind: "copy_output", payload: block.detail.output });
	else if (block.kind === "tool" && block.body) actions.push({ kind: "copy_output", payload: block.body });
	if (ctx.canSend && block.kind === "prompt" && block.body && !ctx.turnInFlight) actions.push({ kind: "rerun", payload: block.body });
	if (ctx.mode === "chat" && ctx.capabilities.includes("rollback") && !ctx.turnInFlight && block.turnId && ctx.rollbackableTurnIds.includes(block.turnId)) {
		actions.push({ kind: "rewind", turnId: block.turnId });
	}
	return actions;
}

export function blockCopyText(block: SessionBlock): string {
	const display = blockDisplay(block);
	const rendered = [display.displayName, display.summary, display.errorText].filter((part): part is string => part !== undefined && part !== "").join("\n").trimEnd();
	const children = block.children ?? [];
	if (children.length === 0) return rendered;
	return [rendered, ...children.map((child) => blockCopyText(child).split("\n").map((line) => `  ${line}`).join("\n"))].join("\n\n");
}

export function blocksToText(blocks: readonly SessionBlock[]): string {
	return blocks.map(blockCopyText).join("\n\n");
}

File: packages/mobile/lib/feature/blocks/logic/block_actions.dart
import 'package:equatable/equatable.dart';
import 'package:operator_mobile/feature/blocks/logic/session_block.dart';
import 'package:operator_mobile/feature/chat/data/model/conversation_snapshot_model.dart';
import 'package:operator_mobile/feature/chat/data/model/conversation_turn_model.dart';

enum BlockActionKind { copyBlock, copyCommand, copyOutput, rerun, rewind }

class BlockActionContext extends Equatable {
  const BlockActionContext({
    required this.mode,
    this.capabilities = const [],
    this.canSend = false,
    this.turnInFlight = false,
    this.rollbackableTurnIds = const [],
  });

  final String mode;
  final List<String> capabilities;
  final bool canSend;
  final bool turnInFlight;
  final List<String> rollbackableTurnIds;

  @override
  List<Object?> get props => [mode, capabilities, canSend, turnInFlight, rollbackableTurnIds];
}

class BlockAction extends Equatable {
  const BlockAction({required this.kind, this.payload, this.turnId});

  final BlockActionKind kind;
  final String? payload;
  final String? turnId;

  @override
  List<Object?> get props => [kind, payload, turnId];
}

bool _turnIsRollbackable(ConversationTurnModel turn) {
  if (turn.id == null || turn.id!.isEmpty) return false;
  if (turn.state == 'running' || turn.state == 'queued') return false;
  if (turn.rolledBack == true) return false;
  if (turn.providerTurnId == null || turn.providerTurnId!.isEmpty) return false;
  return true;
}

List<String> rollbackableTurnIds(ConversationSnapshotModel? snapshot) {
  if (snapshot == null) return const [];
  return [
    for (final turn in snapshot.turns)
      if (_turnIsRollbackable(turn)) turn.id!,
  ];
}

sealed class BlockActions {
  static List<BlockAction> forBlock(SessionBlock block, BlockActionContext ctx) {
    final actions = <BlockAction>[BlockAction(kind: BlockActionKind.copyBlock, payload: copyText(block))];
    final detail = block.detail;
    if (detail is ShellBlockDetail && detail.command != null && detail.command!.isNotEmpty) {
      actions.add(BlockAction(kind: BlockActionKind.copyCommand, payload: detail.command));
    }
    if (detail is ShellBlockDetail && detail.output != null && detail.output!.isNotEmpty) {
      actions.add(BlockAction(kind: BlockActionKind.copyOutput, payload: detail.output));
    } else if (block.kind == BlockKind.tool && (block.result ?? '').isNotEmpty) {
      actions.add(BlockAction(kind: BlockActionKind.copyOutput, payload: block.result));
    } else if (block.kind == BlockKind.tool && block.body.isNotEmpty) {
      actions.add(BlockAction(kind: BlockActionKind.copyOutput, payload: block.body));
    }
    if (ctx.canSend && block.kind == BlockKind.prompt && block.body.isNotEmpty && !ctx.turnInFlight) {
      actions.add(BlockAction(kind: BlockActionKind.rerun, payload: block.body));
    }
    if (ctx.mode == 'chat' &&
        ctx.capabilities.contains('rollback') &&
        !ctx.turnInFlight &&
        block.turnId != null &&
        block.turnId!.isNotEmpty &&
        ctx.rollbackableTurnIds.contains(block.turnId)) {
      actions.add(BlockAction(kind: BlockActionKind.rewind, turnId: block.turnId));
    }
    return actions;
  }

  static String copyText(SessionBlock block) {
    final display = blockDisplay(block);
    final rendered = [display.displayName, display.summary, display.errorText]
        .whereType<String>()
        .where((part) => part.isNotEmpty)
        .join('\n')
        .trimRight();
    final result = block.result ?? '';
    final withResult = result.isEmpty ? rendered : '$rendered\n\n$result';
    final children = block.children ?? const <SessionBlock>[];
    if (children.isEmpty) return withResult;
    return [
      withResult,
      ...children.map((child) => copyText(child).split('\n').map((line) => '  $line').join('\n')),
    ].join('\n\n');
  }

  static String blocksToText(List<SessionBlock> blocks) => blocks.map(copyText).join('\n\n');
}
